src/util/simplify_expr_boolean.cpp

```diff
@@ -17,42 +17,45 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "namespace.h"
 #include "std_expr.h"
 
-bool simplify_exprt::simplify_boolean(exprt &expr)
+simplify_exprt::resultt<> simplify_exprt::simplify_boolean(const exprt &expr)
 {
   if(!expr.has_operands())
-    return true;
-
-  exprt::operandst &operands=expr.operands();
+    return unchanged(expr);
 
   if(expr.type().id()!=ID_bool)
-    return true;
+    return unchanged(expr);
 
   if(expr.id()==ID_implies)
   {
-    if(operands.size()!=2 ||
-       operands.front().type().id()!=ID_bool ||
-       operands.back().type().id()!=ID_bool)
-      return true;
+    if(
+      expr.operands().size() != 2 ||
+      expr.operands().front().type().id() != ID_bool ||
+      expr.operands().back().type().id() != ID_bool)
```

Comment:
Could we use `to_implies_expr` and make sure that does the same amount of sanity checking?

Reply:
added as separate commit.

I wouldn't add the operand type checks to to_implies(). This is what the validate() methods do.